aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -169,8 +180,17 @@ SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
   AT_ASSERT(t.is_sparse());
 
   if (is_same_tensor(r, t)) {
+    // Can't divide an uncoalesced integral tensor accurately. e.g. for a sparse int tensor with value 6
+    // represented as values=[3, 3], integer division by 2 would give values=[1, 1] => 2 instead
+    // of 6 / 2 => 3
+    if (!r.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
+      coalesce_(r);
+    }
     r._values().div_(value);
   } else {
+    if (!t.is_coalesced() && isIntegralType(r.scalar_type(), true)) {
```

Comment:
Right now in the code there are two coalesce checks and calls to in-place coalesce. We can move this to before checking `if_same_tensor(r, t)` so that we only have to check that `t` is coalesced in one location

Furthermore: shouldn't this be a `isIntegralType(t.scalar_type(), true)` (t instead of r), especially since there isn't type promotion in this patch yet?

Reply:
If we move it before the `if (is_same_tensor..` then we have to always coalesce `t` and not `r`. Then the reader has to understand that in the case where we return `r` we're relying on having in-place coalesced an alias of it. It's less code I guess and I suppose that also means the comment is more targeted but are you confident that's less confusing to read?

> shouldn't this be a...

It's actually still `r`. In most cases r.scalar_type==t.scalar_type, but in the cases where it doesn't it would be because someone passed r as `out=`. In that case, if r is floating, we probably don't need to take the perf hit of the in-place coalesce.
```
>>> s=torch.ones([3,3], dtype=torch.int).to_sparse()
>>> t=torch.ones([3,3]).to_sparse()
>>> torch.div(s, 2, out=t)
tensor(indices=tensor([[0, 0, 0, 1, 1, 1, 2, 2, 2],
                       [0, 1, 2, 0, 1, 2, 0, 1, 2]]),
       values=tensor([0.5000, 0.5000, 0.5000, 0.5000, 0.5000, 0.5000, 0.5000,
                      0.5000, 0.5000]),
       size=(3, 3), nnz=9, layout=torch.sparse_coo)
```

Reply:
> It's actually still r. In most cases r.scalar_type==t.scalar_type, but in the cases where it doesn't it would be because someone passed r as out=. In that case, if r is floating, we probably don't need to take the perf hit of the in-place coalesce.

Oh, I see, thanks for explaining that. I didn't realize we converted values() to the correct type before division if the out tensor has another type.

Reply:
> If we move it before the if (is_same_tensor.. then we have to always coalesce t and not r. Then the reader has to understand that in the case where we return r we're relying on having in-place coalesced an alias of it. 

The case that you're describing is the in-place case where `tensor.div_(foo)` calls `div_out(tensor, tensor, foo)`. This is a common paradigm in our codebase: when we implement an in-place function and there already exists an out variant, we call the out-variant with `result` and `input` equal to the tensor, so I don't think this makes the code harder to read.